### PR TITLE
Adding Specifications for Peering

### PR DESCRIPTION
**Pull request by swetha-anantharajaiah**
Here is the link to the Previous PR  https://github.com/Azure/azure-rest-api-specs-pr/pull/575 which was raised after the review for introducing a new Resource Provide Microsoft.Peering

Information about Microsoft Peering - https://peering.azurewebsites.net/peering
User scenarios for Microsoft Peering - http://peering-lab/en-us/azure/peering/peering-howto-directpeering-arm.html

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

File: specification/peering/resource-manager/readme.md
Comment (veronicagg) [inline specification/peering/resource-manager/readme.md]: please add lines for all languages with their corresponding readme files.

**Review by veronicagg — CHANGES_REQUESTED** [review]
This appears to be a new RP, have you done ARM review already? Thanks!

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (nschonni) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: avaiable -> available

**Comment by psanapathi** [comment]
Information about Microsoft Peering - https://peering.azurewebsites.net/peering
User scenarios for Microsoft Peering - http://peering-lab/en-us/azure/peering/peering-howto-directpeering-arm.html

**Comment by veronicagg** [comment]
@psanapathi Have you reviewed the APIs with ARM team? I've labeled the PR to wait for them, but haven't seen a reply. If you need additional info to follow up with them offline, feel free to let me know. Thanks!

**Comment by swetha1microsoft** [comment]
> @psanapathi Have you reviewed the APIs with ARM team? I've labeled the PR to wait for them, but haven't seen a reply. If you need additional info to follow up with them offline, feel free to let me know. Thanks!

Yes, We are following up with @sanjaiganesh offline and we are waiting for him to review the PR

**Comment by veronicagg** [comment]
@shahabhijeet could you please take a look at the .Net SDK generation - it's failing due to: "c:\jenkins\workspace\DotNetSwaggerToSdkCI\scripts\stostools\SToS.proj(17,9): error : Exception: Cannot handle new RPs. New RP(s) detected". How can we help resolving the issue?

**Comment by shahabhijeet** [comment]
startbuild

**Comment by shahabhijeet** [comment]
startbuild

**Comment by adxsdknet** [comment]
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/-1 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/-1/commits/

**Comment by adxsdknet** [comment]
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/-1 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/-1/commits/

**Comment by shahabhijeet** [comment]
@veronicagg and @swetha1microsoft this PR is good to go from .NET SDK point of view.
Sorry for the delay.

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (sanjaiganesh) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: >null [](start = 26, length = 4)

List/collection call need to support paging -  "nextLinkName": "nextLink"

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/examples/ListPeeringOperations.json
Comment (sanjaiganesh) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/examples/ListPeeringOperations.json]: >value [](start = 9, length = 5)

RPs must support 'register' action

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (sanjaiganesh) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: >get [](start = 65, length = 3)

remove 'get' from resource name

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (sanjaiganesh) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: seems collection call.. paging support ?

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (sanjaiganesh) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: >updat [](start = 65, length = 5)

This pattern is not recommended. Also, are you going to merge or replace with the peerInfo list being passed ?
Add 'PUT PeerInfo/{peerInfoName}' instead

**Comment by veronicagg** [comment]
@sarangan12 Peering is a new service, therefore Ruby is failing, please create the necessary folders as required, so the SDK can be produced after the PR is merged. Thanks!

**Comment by veronicagg** [comment]
@swetha1microsoft please ignore Ruby failure for the PR, it's a known issue on our side, due to being a new service. thanks.

**Comment by dsgouda** [comment]
startbuild

**Comment by veronicagg** [comment]
@swetha1microsoft please let me know when you're done reviewing the comments, so I can take a final look. Thanks!

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (sanjaiganesh) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: >peeringLocation [](start = 21, length = 15)

Is this the peering location itself (ex: Redmond) or peeringLocation resource name ? (peeringLocationFoo that represents Redmond,WA). 
I would consider taking peering location resource uri, as part of request body. For ex: names can clash.. just peeringLocation=Redmond is ambiguous it can be present in multiple states/countries)
Does that make sense ?

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (sanjaiganesh) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: >peeringLocation [](start = 9, length = 15)

<same comment>

Is this the peering location itself (ex: Redmond) or peeringLocation resource name ? (peeringLocationFoo that represents Redmond,WA). 
I would consider treating it as a peering location resource uri.
For ex: names can clash.. just peeringLocation=Redmond is ambiguous it can be present in multiple states/countries)
Does that make sense ?

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (sanjaiganesh) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: >put [](start = 7, length = 3)

this is not marked as long running. So, I would presume this api (PUT) completes synchronously.  can you confirm ?

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (sanjaiganesh) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: >properties [](start = 9, length = 10)

Not sure whether we would like to mark properties as required !

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (sanjaiganesh) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: is there a schema difference of the resource , based on the Kind ? 
If yes, then ensure to define the peeringLocation resource schema using 'discriminator' pattern.. Search for x-ms-discriminator-value in the repo.

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: @sanjaiganesh The peeringLocation is a proxy resource, but yes it is the name (Redmond). They are unique and represent the edge locations we have in Azure. It is also unique based on the facility ID which is in the connection, we have a check to ensure the ID and location match.

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: based on the kind direct attributes will be null. Direct Kind means the PeeringDirectProperties cannot be null and exchange must be null.

File: specification/peering/resource-manager/readme.md
Comment (rileymckenna) [reply specification/peering/resource-manager/readme.md]: updated in recent commits.

**Review by rileymckenna — COMMENTED** [review]
changes added in Readme

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: removed the required.

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: we fixed this. you were correct.

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: @sanjaiganesh does this work for you?

Comment (sanjaiganesh) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: It is not ideal. Please check discriminator pattern that I suggested. It will avoid having this additional layer (  "Direct" :  {}  and "Exchange" :  {})


---
In reply to: [266659045](https://github.com/Azure/azure-rest-api-specs/pull/5343#discussion_r266659045) [](ancestors = 266659045)

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (psanapathi) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: Yes, this operation completes synchronously with ProvisioningState "Succeeded". We track the actual peering state under a different variable -- "ConnectionState" -- within the properties.

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (psanapathi) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: The peeringLocation resource "name" and  the "peeringLocation" itself have the same value.

**Comment by adxsdknet** [comment]
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/5498 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/5498/commits/473524cb4185757f4d59d9cd451a931d59395784

**Comment by veronicagg** [comment]
@swetha1microsoft when you're done with the API review, please let me know and I'll take a final look. Thanks!

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (veronicagg) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: is there any documentation that could be added on what values location should take?

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (veronicagg) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: is this readonly?

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (veronicagg) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: typo: pee's ASN -> peer's ASN :)

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (veronicagg) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: would you want to abstract the "resource" properties into a "Resource" definition and use it with "allOf" for ["Peering"](https://github.com/Azure/azure-rest-api-specs/pull/5343/files#diff-9dd14f9f472a0f883223a1a597b3be79R669) and this one? - see https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v1/types.json

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (veronicagg) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: should it have a more descriptive description?

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (veronicagg) [inline specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: is this also a "resource" shoud it share the "resource" definition with "Peer" and should it be tagged x-ms-azurre-resouce?

**Review by veronicagg — CHANGES_REQUESTED** [review]
I left a few more comments inline. Thanks!

File: specification/peering/resource-manager/readme.typescript.md
Comment (rileymckenna) [reply specification/peering/resource-manager/readme.typescript.md]: complete

File: specification/peering/resource-manager/readme.typescript.md
Comment (rileymckenna) [reply specification/peering/resource-manager/readme.typescript.md]: complete

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: yes, we do. We expect that users will Get the avaliable peering locations first then use one the listed. Its somewhere around 200 peering locations and adding more each year.

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: type object is readonly

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: complete

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: no it is a proxy read only resource.

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: give me a second to talk it out with our team.

Comment (rileymckenna) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: we made these changes along with the PeeringLocation

**Comment by rileymckenna** [comment]
@veronicagg completed all changes and pushed the latest

File: specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json
Comment (veronicagg) [reply specification/peering/resource-manager/Microsoft.Peering/preview/2019-03-01-preview/peering.json]: ok, you could evaluate updating your description for customers to better make use of it.

**Review by veronicagg — APPROVED** [review]
Looks good, thanks!

**Pull request merged**
